tests/smoke/alp_zgeqrf.cpp

```diff
@@ -0,0 +1,301 @@
+template<
+	typename T
+>
+std::vector< T > generate_rectangular_matrix_data(
+	size_t N,
+	size_t M,
+	const typename std::enable_if<
+		grb::utils::is_complex< T >::value,
+		void
+	>::type * const = nullptr
+) {
+	std::vector< T > data( N * M );
+	std::srand( RNDSEED );
+	for( size_t i = 0; i < N; ++i ) {
+		for( size_t j = 0; j < M; ++j ) {
+			T val( std::rand(), std::rand() );
+			data[ i * M + j ] = val / std::abs( val );
+		}
+	}
+	return data;
+}
+
+//** generate random rectangular matrix data: real version */
+template<
+	typename T
+>
+std::vector< T >  generate_rectangular_matrix_data(
+	size_t N,
+	size_t M,
+	const typename std::enable_if<
+		!grb::utils::is_complex< T >::value,
+		void
+	>::type * const = nullptr
+) {
+	std::vector< T > data( N * M );
+	std::srand( RNDSEED );
+	for( size_t i = 0; i < N; ++i ) {
+		for( size_t j = 0; j < M; ++j ) {
+			//data[ k ] = static_cast< T >( i + j*j ); // easily reproducible
+			data[ i * M + j ] = static_cast< T >( std::rand() )  / RAND_MAX;
+		}
+	}
+	return data;
+}
+
+//** check if rows/columns or matrix Q are orthogonal */
+template<
+	typename T,
+	typename Structure,
+	typename ViewType,
+	class Ring = Semiring< operators::add< T >, operators::mul< T >, identities::zero, identities::one >
+>
+RC check_overlap( alp::Matrix< T, Structure, alp::Density::Dense, ViewType > &Q, const Ring & ring = Ring() ) {
+	// more elegent would be check mxm(Q,conjugate(transpose(Q))) == identity
+	RC rc = SUCCESS;
+	const size_t n = nrows( Q );
+#ifdef DEBUG
+	std::cout << "Overlap matrix for Q:\n";
+#endif
+	for ( size_t i = 0; i < n; ++i ) {
+		auto vi = get_view( Q, i, utils::range( 0, n ) );
+		for ( size_t j = 0; j < n; ++j ) {
+			auto vj = get_view( Q, j, utils::range( 0, n ) );
+			Scalar< T > alpha( ring.template getZero< T >() );
+			rc = dot( alpha, vi, vj, ring );
+			if( rc != SUCCESS ) {
+				std::cerr << " dot( alpha, vi, vj, ring ) failed\n";
+				return PANIC;
+			}
+			if( i == j ) {
+				if( std::abs( *alpha - ring.template getOne< T >() ) > tol ) {
+					std::cerr << " vector " << i << " not normalized\n";
+					return PANIC;
+				}
+			} else {
+				if( std::abs( *alpha ) > tol ) {
+					std::cerr << " vector " << i << " and vctor " << j << " are note orthogonal\n";
+					return PANIC;
+				}
+			}
+#ifdef DEBUG
+			std::cout << "\t" << std::abs( *alpha );
+#endif
+		}
+#ifdef DEBUG
+		std::cout << "\n";
+#endif
+	}
+#ifdef DEBUG
+	std::cout << "\n";
+#endif
+	return rc;
+}
+
+
+//** check solution by calculating H-QR */
+template<
+	typename D,
+	typename StructureGen,
+	typename StructureOrth,
+	class Minus = operators::subtract< D >,
+	class Ring = Semiring< operators::add< D >, operators::mul< D >, identities::zero, identities::one >
+>
+RC check_solution(
+	alp::Matrix< D, StructureGen, alp::Density::Dense > &H,
+	alp::Matrix< D, StructureOrth, alp::Density::Dense > &Q,
+	alp::Matrix< D, StructureGen, alp::Density::Dense > &R,
```

Comment:
Similar comment as above: View, ImfR, ImfC and backend should be template parameters.